src/mon/FSCommands.cc

```diff
@@ -268,6 +287,41 @@ class FsNewHandler : public FileSystemCommandHandler
       mon->osdmon()->wait_for_writeable(op, new PaxosService::C_RetryMessage(mon->mdsmon(), op));
       return -EAGAIN;
     }
+
+    bool recover = false;
+    cmd_getval(cmdmap, "recover", recover);
+
+    auto&& fs = fsmap.create_filesystem(fs_name, metadata, data,
+        mon->get_quorum_con_features(), fscid, recover);
+
+    // set fs options
+    int ret = 0;
+    string set_fsops_info;
+    int max_mds = 1;
+    for(int i = 1 ; i < (int)fsops_vec.size() ; i+=2) {
+      std::ostringstream oss;
+      ret = set_val(mon, fsmap, op, cmdmap, oss, fs_name, fsops_vec[i],
+                    fsops_vec[i+1]);
+      if(ret < 0) {
+        ss << oss.str();
+        break;
+      }
+      if((i + 2) <= (int)fsops_vec.size()) {
+        set_fsops_info.append("; ");
+      }
+      set_fsops_info.append(oss.str());
+      if(fsops_vec[i] == "max_mds") {
+        max_mds = stoi(fsops_vec[i+1]);
+      }
+    }
+    if (ret < 0) {
+      fsmap.erase_filesystem(fs->fscid);
```

Comment:
`FSMap::create_filesystem` may have other side-effects. In particular, `legacy_client_fscid` and `next_filesystem_id` may be set. You should refactor that code to only set `legacy_client_fscid` (via a method) only if `fs new` succeeds. Or better would be change `FSMap::create_filesystem` to only create the fs and then another method to "add" it (or "commit", or whatever).

Reply:
`legacy_client_fscid` is set only if it's the first fs:
```
if (filesystems.size() == 1) {
    legacy_client_fscid = fs->fscid;
  }
```

So, with a cluster with no fs, i checked `legacy_client_fscid` before and after running `fs new` with invalid set op; it's value remains `-1`.

For `next_filesystem_id`, yeah it needs to rolled back to val before incrementing it

Reply:
done, PTAL

Reply:
> So, with a cluster with no fs, i checked legacy_client_fscid before and after running fs new with invalid set op; it's value remains -1.

That doesn't make sense to me. It should be changed. There must not be side-effects when unwinding an operation. So no matter what, the old value must be restored.

I strongly recommend a create/commit style interface. Having methods to individually unwind settings is cumbersome, error-prone, and adds noise to the API.

Reply:
The thing is `set_val()` cannot take place until the fs we (try to)create is in `filesystems`, addition of fs takes place like this `filesystems[fs->fscid] = fs;`, so we need an fscid while inserting. So let's say we have  `create_filesystem()` split into `create()` and `commit()`, we cannot set fs ops after `create()`.

An approach I was thinking is: creating a temp map to store fs there first, with a dummy fscid (or maybe some other way) -> run `set_val()` on it -> went fine? call `commit()` where we manage the `legacy_client_fscid` and `next_filesystem_id` and add the fs to `filesystems` else clean up the temp map and return. What do you think?

Reply:
> The thing is `set_val()` cannot take place until the fs we (try to)create is in `filesystems`, addition of fs takes place like this `filesystems[fs->fscid] = fs;`, so we need an fscid while inserting. So let's say we have `create_filesystem()` split into `create()` and `commit()`, we cannot set fs ops after `create()`.

This is a solvable problem with some reworking of the interface. The main barrirer here that requires the use of the fsmap is `FSMap::modify_filesystem`. Add another:

```
    template<typename T>
    void modify_filesystem(std::shared_ptr<Filesystem>& fs, T&& fn)
    {
      fn(fs);
      fs->mds_map.epoch = epoch;
    } 
```

Now it no longer needs to be the "filesystems" map (to lookup by fscid). Consequently, your `set_val` method should look like:

```
  int FileSystemCommandHandler::set_val(Monitor *mon, FSMap& fsmap, MonOpRequestRef op,
              const cmdmap_t& cmdmap, std::ostream &ss, std::shared_ptr<Filesystem>& fs,
              std::string var, std::string val)
```

Pass the `fs` handle to it from the `fs new` and `fs set` handlers. From my scan of the code, the last remaining "issue" is setting `allow_standby_replay` and how it will fail standby-replay daemons if set to false. Obviously for a fresh/new fs, this will by default be `false`. However, it can still be set to false explicitly via `set allow_standby_replay false`. Still, no ranks are assigned yet so no existing standby-replay will be removed. But, the code still checks if the osdmap is writable even though no work will be done. You should alter that behavior to only do that check if `!to_fail.empty()`. I would also add a comment to that code for the reader to be mindful that `set_val` may be called on a new but uncommitted file system.

Reply:
Another issue is in `fs set`, `FSMap::get_filesystem` returns a const pointer, I cannot set this new `void modify_filesystem` to use fs as const pointer else it won't let me assign values fs's members. Should I create one a helper to return non-const ref to fs?

Reply:
or instead create `set_val` for `fs set` that accepts `fs_name`

Reply:
I've created a new get_filesystem that returns non-const sp. I was thinking for a way such that i can accommodate returning both const and non-const within same snippet, something like this

```
template<typename T>
T FSMap::get_filesystem(std::string_view name)
{
  std::shared_ptr<Filesystem const> const_fs = nullptr;
  std::shared_ptr<Filesystem> fs = nullptr;
  
  for (const auto& p : filesystems) {
    if (p.second->mds_map.fs_name == name) {
      if(std::is_same<T, Filesystem::ref>::value) {
        fs = p.second;
      } else {
        const_fs = p.second;
      }
    }
  }
}
```

this throws a ton of errors - "no known conversion", "template argument deduction/substitution failed"

Reply:
So `modify_filesystem` exists to protect the programmer from forgetting to increment the epoch. We don't have that problem when first creating the file system.

I think the right thing to do here is to modify your `set_val` method to take `std::variant<std::shared_ptr<Filesystem>,fscid_t>`. See also 

https://en.cppreference.com/w/cpp/utility/variant

There will be a small amount of programming work to run the lambdas normally passed to `FSMap::modify_filesystem` directly on the `Filesystem` object instead. Suggest using a helper function in `FSCommands.cc` to do that.

Reply:
This looks a bit overkill, also making changes to fs from `FSCommands` instead of `FSMap` sounds a bit weird. I'd advocate for the current changes + adding a new `get_filesystem` that accepts fs pointer because:
1) Having a `modify_filesystem` that accepts fs pointer would be helpful in other cases too where it might be needed to modify fs using pointer.
2) We already have `get_filesystem` methods that return non-const pointer, so i think adding one more should not be an issue.
3) Keeping fs related changes to `FSMap` looks correct, clean and modular.

What do you think?

Reply:
eliminated the new `get_filesystem` by getting it using `fscid` in `SetHandler`